### PR TITLE
Track test events

### PR DESCRIPTION
**Pull request by jpjpjp**
This is primarily a big rewrite of the tests along with a few little bug fixes I made as I was doing them.    This PR is designed to be released together with https://github.com/WebexCommunity/webex-node-bot-framework/pull/96 (which I just merged) which is the one where I added the new fields to the trigger object that @adamweeks requested last year.

Would be great of have a second set of eyes on this if anyone has time, but if not I'll likely merge myself in the next few days.

## v 2.5.0
* Added the new `command` and `prompt` elements to the `trigger` object that is passed as a parameter to matching `framework.hears()` callbacks.
  * `trigger.command` will be the portion of the user message that matched the phrase specified in the `framework.hears()`
  * `trigger.prompt` will be words in the message that follow the command (minus any bot mention).
* Clarified `bot.reply()` docs.  This method supports a reply to a reply if the replyTo parameter is a message object.
* Updated from EventEmitter to EventEmitter2 class. This was done to support new tests, but if desired it now allows developers to write catch all event handlers in the form:
```js
framework.onAny((eventName, ...args) => {
  if (eventName == 'log') {
    console.log(`Got a framework log event: "${args[0]}"`);
    return;
  } else {
    // Handle other events as desired
  }
}
```
* This release includes a major overhaul of the tests in the hopes that it is easier for contributors to understand, run, and create new tests.   Improvements include:
  * Tests fail if unexpected events occur while waiting for expected events, with an error message describing what event was unexpected.  This uses the new `framework.onAny()` capability.  See the `registerUnexpectedEventsHandler()` in [test/common/common.js](../test/common/common.js)
  * When tests timeout while waiting for expected events, details are now provided on which events were expected, which arrived, and which were missing.
  * Better code reuse and [documentation](../docs/tests.md).

### CODE REVIEW

File: docs/header.md [hunk 1]
```diff
@@ -23,6 +23,8 @@ framework.hears('echo', (bot, trigger) => {
 }, '**echo** - I\'ll echo back the rest of your message');
 ```
 
+This release also includes updated [Contribution Guidelines](./docs/contributing.md) including details on [running the tests](./tests.md)
```
Comment (jeremywillans) [inline docs/header.md:26]: contributing doc hasnt been updated?

File: docs/tests.md [hunk 2]
```diff
@@ -182,3 +185,44 @@ While developers are encouraged to generate the tokens required for the tests us
...
+* `testInfo.config` is set up by the calling test function primarily to tell the class about the test user, bot instance, and space instance that will be active for a given test.   Typically, the key elements of the `testInfo.config` object, namely the `botUnderTest`, `userUnderTest`, and `roomUnderTest`, elements are set automatically by the common `before()` and `after()` handlers in the [before-after-user-created-room.js](../test/common/before-after-user-created-room.js) and [before-after-bot-created-room.js](../test/common/before-after-bot-created-room.js)
+* `testInfo.in` is an object that is set up by the common wrapper functions as it evaluates how the request should behave based on the framework configuration and test state.  Test code outside of the common class implementation should not manipulate the `testInfo.in` class but it can be helpful to inspect it when tests fail to determine what went wrong.
+* `testInfo.out` is an object that is setup and populated by the common wrapper functions as API calls return and events are processed.   Generally the tests will look to populate the `out` class so that it matches the `in` class.
+
+## When to add tests
```
Comment (jeremywillans) [inline docs/tests.md:218]: is it worth putting/linking this in the contributing guide?

File: requirements.txt
Comment (jeremywillans) [inline requirements.txt:1]: is this file still needed?

File: docs/header.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpjpjp) [reply docs/header.md:26]: Contribution Guidelines were updated fairly recently...I'm putting it into the version history here simply to try to get people to take a look.    I wanted to wait until the test documentation was solid.

File: .clj-kondo/.cache/v1/lock
Comment (jeremywillans) [inline .clj-kondo/.cache/v1/lock:1]: maybe update .gitignore to remove this?

File: .lsp/.cache/db.transit.json
Comment (jeremywillans) [inline .lsp/.cache/db.transit.json:1]: update .gitignore

File: lib/membership-rules.js [hunk 3]
```diff
@@ -535,4 +535,4 @@ function isDomainAllowed(domain, restrictedDomainsList) {
   } else {
     return false;
   }
-};
\ No newline at end of file
+}
```
Comment (jeremywillans) [inline lib/membership-rules.js:538]: missing EOF LB?

**Review by jeremywillans — APPROVED** [review]
LGTM!

File: lib/membership-rules.js (hunk 3 above)
Comment (jpjpjp) [reply lib/membership-rules.js]: Done
